manticore/platforms/evm.py

```diff
@@ -915,19 +989,45 @@ def setstate(state, value):
                              policy='ALL')
         try:
             self._check_jmpdest()
-            last_pc, last_gas, instruction, arguments = self._checkpoint()
+            last_pc, last_gas, instruction, arguments, fee, allocated = self._checkpoint()
+            self._consume(fee)
             result = self._handler(*arguments)
             self._advance(result)
-        except ConcretizeStack as ex:
-            self._rollback()
-            pos = -ex.pos
 
+        except ConcretizeGas as ex:
+            def setstate(state, value):
+                state.platform.current._gas = value
+            raise Concretize("Concretize gas",
+                             expression=self._gas,
+                             setstate=setstate,
+                             policy='MINMAX')
+        except ConcretizeFee as ex:
             def setstate(state, value):
-                self.stack[pos] = value
-            raise Concretize("Concretize Stack Variable",
-                             expression=self.stack[pos],
+                current_vm = state.platform.current_vm
+                _pc, _old_gas, _instruction, _arguments, _fee, _allocated = current_vm._checkpoint_data
+                current_vm._checkpoint_data = (_pc, _old_gas, _instruction, _arguments, _value, _allocated)
+            raise Concretize("Concretize current instruction fee",
+                             expression=fee,
                              setstate=setstate,
                              policy=ex.policy)
+        except ConcretizeArgument as ex:
+            pos = ex.pos - 1
+
+            def setstate(state, value):
+                current_vm = state.platform.current_vm
+                _pc, _old_gas, _instruction, _arguments, _fee, _allocated = current_vm._checkpoint_data
+                new_arguments = []
+                for old_arg in _arguments:
+                    if len(new_arguments) == pos:
+                        new_arguments.append(value)
+                    else:
+                        new_arguments.append(old_arg)
+                    current_vm._checkpoint_data = (_pc, _old_gas, _instruction, new_arguments, _fee, _allocated)
```

Comment:
shouldn't this be out of the `for` loop?
```suggestion
                current_vm._checkpoint_data = (_pc, _old_gas, _instruction, new_arguments, _fee, _allocated)
```